infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespace.java

```diff
@@ -154,6 +157,14 @@ void prepare(boolean canCreate, Map<String, String> labels, Map<String, String>
             format("Creating the namespace '%s' is not allowed, yet it was not found.", name));
       }
       namespace = create(name, client);
+      Secret secret =
```

Comment:
I have 2 questions on this....
  1) Why there is no equivavlent code in Openshift ?
  2) Can we move that secret creation out of this class, please? It breaks it's contract badly, mixing different stuff together.. at least into `KubernetesNamespaceFactory`...

Reply:
> Why there is no equivavlent code in Openshift ?

That's my bad, added the Openshift implementation in the latest commit
>Can we move that secret creation out of this class, please? It breaks it's contract badly, mixing different stuff together.. at least into KubernetesNamespaceFactory...

* I don't understand why it breaks the contract, so the class javadoc says: `Defines an internal API for managing subset of objects inside {@link Namespace} instance.`. why the secret object can't be a part of the subset of objects inside?
* The main reason why I put the secret creation in that method is that because the `prepare` method contains a check for created namespace, e.g. we mustn't create a new credentials secret if the namespace is already exists. In case of moving this logic to  `KubernetesNamespaceFactory` we would need to duplicate the check.
* Something similar already present in the `OpenshiftProject.java`: https://github.com/eclipse-che/che-server/blob/fb0fe45e94186960508349b9f7d41012beeea945/infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/project/OpenShiftProject.java#L143-L157

Reply:
>Defines an internal API for managing subset of objects inside {@link Namespace} instances
Right, provides API, but does not work with that objects directly.

If we consider the program as the set of layers,  this class sits at bottom (DAO)-layer, which works purely with it's own instances (create, labels, annotate etc) and must contain, ideally, no other business logic at all. For business logic there is "Manager" layer, which is exacly `KubernetesNamespaceFactory`.

Reply:
Fixed, moved the logic to Factory classes.